src/compiler/types.ts

```diff
@@ -1639,6 +1652,28 @@ namespace ts {
     export type StringLiteralLike = StringLiteral | NoSubstitutionTemplateLiteral;
     export type PropertyNameLiteral = Identifier | StringLiteralLike | NumericLiteral;
 
+    export interface TemplateTypeNode extends TypeNode {
+        kind: SyntaxKind.TemplateType,
+        readonly head: TemplateHead;
+        readonly templateSpans: NodeArray<TemplateTypeSpan>;
+    }
+
+    export interface TemplateTypeSpan extends TypeNode {
+        readonly kind: SyntaxKind.TemplateTypeSpan,
+        readonly parent: TemplateTypeNode;
+        readonly casing: TemplateCasing;
+        readonly type: TypeNode;
+        readonly literal: TemplateMiddle | TemplateTail;
+    }
+
+    export const enum TemplateCasing {

```

Comment:
```suggestion
    export const enum StringTemplateCasing {
```

I feel weird that this is *part* of the node instead of a first-class operator.

Does this need to be publicly exported?